### PR TITLE
planner: ban tiflash engine when the statement is not read only

### LINKED ISSUE #17851: TiDB doesn't lock non-existent unique-key when select for update from TiFlash
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
create table:
`create table if not exists txn0 (id  int not null primary key, val text); `

create tiflash replica and wait for the replica become available:
`alter table txn0 set tiflash replica 1;`

in the first terminal, type the following sql:
`set @@session.tidb_isolation_read_engines = "tiflash";`
`begin pessimistic;`
`select val from txn0 where id = 25 for update;`

in the second terminal, type the following sql:
`insert into txn0 (id, val) values (25, '5');`

### 2. What did you expect to see? (Required)
The insert statement will get stuck.
### 3. What did you see instead (Required)
The insert statement succeed.
### 4. Affected version (Required)
v4.0.0,v3.1.0

### PR DESCRIPTION
**Pull request by lzmhhh123**
### What problem does this PR solve?

Issue Number: close #17851

Problem Summary: When the SQL is like `select *** for update`. It's a not-read-only statement. The TiFlash plan changes the point get to point range table scan. But the lock operator has different performance between coprocessor and point get. So this causes the wrong results when `select *** for update` by using the TiFlash. We decide to ban the TiFlash engine for a not-read-only statement.

### What is changed and how it works?

What's Changed: ban TiFlash engine for a not-read-only statement.

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### Release note 

- ban the TiFlash engine for a not-read-only statement.

### CODE REVIEW

**Comment by zz-jason** [comment]
please add a unittest

**Comment by lzmhhh123** [comment]
> please add a unittest

Done.

File: planner/optimize.go [hunk 1]
```diff
@@ -32,15 +32,60 @@ import (
...
+	}
+	return nil, plannercore.ErrStmtNotFound
+}
+
+// IsReadOnly check whether the ast.Node is a read only statement.
+func IsReadOnly(node ast.Node, vars *variable.SessionVars) bool {
+	if execStmt, isExecStmt := node.(*ast.ExecuteStmt); isExecStmt {
+		s, err := GetPreparedStmt(execStmt, vars)
+		if err != nil {
+			logutil.BgLogger().Error("GetPreparedStmt failed", zap.Error(err))
```
Comment (XuHuaiyu) [inline planner/optimize.go:66]: I think we can use `Warn` instead of `Error`

File: planner/core/integration_test.go [hunk 2]
```diff
@@ -1154,6 +1154,38 @@ func (s *testIntegrationSuite) TestOptimizeHintOnPartitionTable(c *C) {
...
+	c.Assert(exists, IsTrue)
+	for _, tblInfo := range db.Tables {
+		if tblInfo.Name.L == "t" {
+			tblInfo.TiFlashReplica = &model.TiFlashReplicaInfo{
+				Count:     1,
+				Available: true,
+			}
+		}
+	}
+	err := tk.ExecToErr("select * from t for update")
```
Comment (XuHuaiyu) [inline planner/core/integration_test.go:1177]: It seems that we need to use QueryToErr here

Comment (lzmhhh123) [reply planner/core/integration_test.go:1177]: `QueryToErr` will assert false in the function when checking the results.

**Pull request merged**